assets/app/features/upgrade/steps.factory.spec.js

```diff
@@ -118,5 +406,87 @@ describe('Stepes Factory', function () {
             });
 
         });
+
+        describe('stepByState function', function () {
+            _.forEach(mockedInitialSteps, function (mockedStep) {
+                it('should return step with state=' + mockedStep.state + ' when called', function () {
+                    expect(upgradeStepsFactory.stepByState(mockedStep.state)).toEqual(mockedStep);
+                });
+            });
+
+            it('should return undefined when called with nonexistent state', function () {
+                expect(upgradeStepsFactory.stepByState('--dummy state--')).not.toBeDefined();
+            });
+        });
+
+        describe('stepByID function', function () {
+            _.forEach(mockedInitialSteps, function (mockedStep) {
+                it('should return step with ID=' + mockedStep.id + ' when called', function () {
+                    expect(upgradeStepsFactory.stepByID(mockedStep.id)).toEqual(mockedStep);
+                });
+            });
+
+            it('should return undefined when called with nonexistent ID', function () {
+                expect(upgradeStepsFactory.stepByID(-1)).not.toBeDefined();
+                expect(upgradeStepsFactory.stepByID(9999)).not.toBeDefined();
+            });
+        });
+
+        describe('lastStateForRestore function', function () {
+            describe('when there is no stored last state', function () {
+                beforeEach(function () {
+                    localStorage.removeItem(UPGRADE_LAST_STATE_KEY);
+                });
+                it('should return "upgrade-landing" before any step', function() {
```

Comment:
It would be interesting to have this set of validations be done by iterating through `stepsToStateMap` from the steps.factory and cover all cases, if such map is stored as a public property of the factory or a module's constant... What do you think?